### PR TITLE
fix/15753: suggested address not populate if unknown to google db

### PR DESCRIPTION
**Pull request by tienifr**
### Details

Some suggested places from Google Place Autocomplete do not automatically populate to other fields in Home Address page if they do not exists on Google database. This PR fixes the issue by using the suggested result as fallback values when there's no data on Google database.

### Fixed Issues

$ https://github.com/Expensify/App/issues/15753   
PROPOSAL: https://github.com/Expensify/App/issues/15753#issuecomment-1464833715

### Tests

1. Login with any account
2. Click on Profile avatar > Profile > Personal details > Home address
3. In the Address line 1 field type in: **I**
4. Select **I-35** (i.e. I-35 Ottawa, KS, USA)
5. Verify that other fields are filled automatically just as the suggestion (i.e. Interstate 35 - Ottawa - Kansas - United States)
6. Select **I-80** (i.e. I-80 Seward, NE, USA)
7. Verify that other fields are filled automatically just as the suggestion (i.e. Interstate 80 - Seward - Nebraska - United States)
8. In the Address line 1 field, type in: N12 ireland
9. Select the option N12, Knockaconny, County Monaghan, Ireland
10. Make sure that the City, State / Province, and Country fields get automatically filled
11. In the Address line 1 field, type in: BR-101
12. Select the option that starts with BR-101, Linhares
13. Make sure that the City, State / Province, and Country fields get automatically filled
- [x] Verify that no errors appear in the JS console

### Offline tests

Same above

### QA Steps

1. Login with any account
2. Click on Profile avatar > Profile > Personal details > Home address
3. In the Address line 1 field type in: **I**
4. Select **I-35** (i.e. I-35 Ottawa, KS, USA)
5. Verify that other fields are filled automatically just as the suggestion (i.e. Interstate 35 - Ottawa - Kansas - United States)
6. Select **I-80** (i.e. I-80 Seward, NE, USA)
7. Verify that other fields are filled automatically just as the suggestion (i.e. Interstate 80 - Seward - Nebraska - United States)
8. In the Address line 1 field, type in: N12 ireland
9. Select the option N12, Knockaconny, County Monaghan, Ireland
10. Make sure that the City, State / Province, and Country fields get automatically filled
11. In the Address line 1 field, type in: BR-101
12. Select the option that starts with BR-101, Linhares
13. Make sure that the City, State / Province, and Country fields get automatically filled
- [ ] Verify that no errors appear in the JS console

### PR Author Checklist

- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for the expected offline behavior in the `Offline steps` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [x] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
    - [x] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] Android / native
    - [x] Android / Chrome
    - [x] iOS / native
    - [x] iOS / Safari
    - [x] MacOS / Chrome / Safari
    - [x] MacOS / Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
      - [x] If any non-english text was added/modified, I verified the translation was requested/reviewed in #expensify-open-source and it was approved by an internal Expensify engineer. Link to Slack message:
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is correct English and approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] I verified that if a function's arguments changed that all usages have also been updated correctly
- [x] If a new component is created I verified that:
    - [x] A similar component doesn't exist in the codebase
    - [x] All props are defined accurately and each prop has a `/** comment above it */`
    - [x] The file is named correctly
    - [x] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [x] The only data being stored in the state is data necessary for rendering and nothing else
    - [x] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [x] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [x] All JSX used for rendering exists in the render method
    - [x] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(themeColors.componentBG`)
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.
- [x] I have checked off every checkbox in the PR author checklist, including those that don't apply to this PR.

### Screenshots/Videos
<details>
<summary>Web</summary>

[Screencast from 21-03-2023 23:46:11.webm](https://user-images.githubusercontent.com/113963320/226681846-acb66668-9885-4441-9e88-c4bcb0128b2a.webm)

</details>

<details>
<summary>Mobile Web - Chrome</summary>

[Screencast](https://user-images.githubusercontent.com/113963320/226682696-fa4a3f69-2e30-4ae0-bfd1-623fb269257c.mp4)

</details>

<details>
<summary>Mobile Web - Safari</summary>

https://user-images.githubusercontent.com/113963320/226799244-f286a4b6-77f1-45f0-889e-db92e011cd43.mp4

</details>

<details>
<summary>Desktop</summary>

https://user-images.githubusercontent.com/113963320/226798047-16b054bc-685c-486f-8972-b9fe50889f88.mp4

</details>

<details>
<summary>iOS</summary>

https://user-images.githubusercontent.com/113963320/226798035-fe095250-8a5f-45f0-b91b-8f4bacc207f4.mp4

</details>

<details>
<summary>Android</summary>

https://user-images.githubusercontent.com/113963320/226686354-e036c2a1-7de9-46f7-8938-53dea06b9a99.mp4

</details>

### CODE REVIEW

**Comment by tienifr** [comment]
Hi @sobitneupane Please help to review this PR. Many thanks

**Review by sobitneupane — COMMENTED** [review]
@tienifr We had discussed on using fallback solution only if the predicted country is USA. It looks like we are using fallback solution for all countries. Will the solution work for all countries?

File: src/libs/GooglePlacesUtils.js [hunk 1]
```diff
@@ -27,7 +27,25 @@ function getAddressComponents(addressComponents, fieldsToExtract) {
...
+/**
+ * Finds an address term by type, and returns the value associated to key. Note that each term in the address must
+ * conform to the following ORDER: <street, city, state, country>
+ *
+ * @param {Array} addressTerms
+ * @returns {Object}
+ */
+function getPlaceAutocompleteTerms(addressTerms) {
+    const fieldsToExtract = ['country', 'state', 'city', 'street'];
+    const result = _.map(fieldsToExtract, () => '');
```
Comment (sobitneupane) [inline src/libs/GooglePlacesUtils.js:39]:
```suggestion
    const result = {};
```

**Comment by tienifr** [comment]
@francoisl Take a look at the PR when you're available. Thanks!

**Review by sobitneupane — COMMENTED** [review]
Some Issues on other countries:
Screenshot 2023-03-24 at 18 50 24

**Comment by tienifr** [comment]
@sobitneupane There are 2 issues there:
1. The city field is filled with `false`
2. The Zip/Postal code is missing

The first issue is my fault due to the way I get the term within `getPlaceAutocompleteTerms`, I've fixed this.

The second issue is not related to this PR.

**Comment by sobitneupane** [comment]
@francoisl I think we should use this fallback solution only for USA. Address formats for other countries are usually different and might not work always. What do you suggest?

**Comment by tienifr** [comment]
@francoisl Are you available?

**Comment by tienifr** [comment]
@arielgreen Do you think we need to assign someone else to review this PR? As I think @francoisl may be OOO.

**Comment by francoisl** [comment]
Hi, sorry I was out of office but I'm back now.

> I think we should use this fallback solution only for USA. Address formats for other countries are usually different and might not work always.

My rationale for proposing _not_ to restrict to USA was that we might as well try to use the fallback logic for all countries, on the off chance that the `terms` follow the same `'street', 'city', 'state', 'country'` order.

The difficult part of testing is that it's hard to find addresses similar to `I-80` of the original issue for which Google Maps' API doesn't return details, but for other countries. I found a couple examples by just entering random road names I found on Google Maps:
- Search for `N12 Ireland`, choose the first suggestion that contains `Knockaconny`: with this proposed fix, we'd correctly fill in the city
- Search for `BR-101`, choose the first suggestion that contains `Linhares` from Brazil: with this fix, we'd also correctly fill in the city

And so on, there must be other examples in other countries, but it's time-consuming to search for them.

----

Also @tienifr there is a conflict on the branch, do you mind updating please?

**Comment by tienifr** [comment]
@francoisl I've already resolved it. Btw, I'm not sure your final conclusion on whether to apply this logic to the US only.

**Comment by francoisl** [comment]
> Btw, I'm not sure your final conclusion on whether to apply this logic to the US only.

Hm sorry, what do you mean? Are you rethinking your solution and saying we should restrict the fix to the US only? I'm still thinking we should not restrict the fix to the US only, as suggested by my examples above.

We can even add those two examples to the test and QA steps, for example:

[... current steps]
8. In the Address line 1 field, type in: `N12 ireland`
9. Select the option `N12, Knockaconny, County Monaghan, Ireland`
10. Make sure that the `City`, `State / Province`, and `Country` fields get automatically filled
11. In the Address line 1 field, type in: `BR-101`
12. Select the option that starts with `BR-101, Linhares`
13. Make sure that the `City`, `State / Province`, and `Country` fields get automatically filled

**Review by francoisl — APPROVED** [review]

**Comment by tienifr** [comment]
@francoisl Thanks for your comments, I just updated the test step. @sobitneupane bump

**Review by sobitneupane — COMMENTED** [review]
### Screenshots/Videos
<details>
<summary>Web</summary>
Screenshot 2023-04-01 at 08 05 49
Screenshot 2023-04-01 at 08 05 32
Screenshot 2023-04-01 at 08 05 08
Screenshot 2023-04-01 at 08 04 29
Screenshot 2023-04-01 at 08 03 16
Screenshot 2023-04-01 at 08 07 14

</details>


<details>
<summary>Mobile Web - Chrome</summary>

Screenshot 2023-04-01 at 08 15 57
Screenshot 2023-04-01 at 08 15 26
Screenshot 2023-04-01 at 08 15 07




</details>

<details>
<summary>Mobile Web - Safari</summary>







</details>

<details>
<summary>Desktop</summary>

Screenshot 2023-04-01 at 08 08 50
Screenshot 2023-04-01 at 08 08 42
Screenshot 2023-04-01 at 08 09 40
Screenshot 2023-04-01 at 08 09 30
Screenshot 2023-04-01 at 08 09 04
Screenshot 2023-04-01 at 08 11 37

</details>

<details>
<summary>iOS</summary>





</details>

<details>
<summary>Android</summary>






</details>

**Review by sobitneupane — APPROVED** [review]
## Reviewer Checklist

- [x] I have verified the author checklist is complete (all boxes are checked off).
- [x] I verified the correct issue is linked in the `### Fixed Issues` section above
- [x] I verified testing steps are clear and they cover the changes made in this PR
    - [x] I verified the steps for local testing are in the `Tests` section
    - [x] I verified the steps for Staging and/or Production testing are in the `QA steps` section
    - [x] I verified the steps cover any possible failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
- [x] I checked that screenshots or videos are included for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I verified tests pass on **all platforms** & I tested again on:
    - [x] Android / native
    - [x] Android / Chrome
    - [x] iOS / native
    - [x] iOS / Safari
    - [x] MacOS / Chrome / Safari
    - [x] MacOS / Desktop
- [x] If there are any errors in the console that are unrelated to this PR, I either fixed them (preferred) or linked to where I reported them in Slack
- [x] I verified proper code patterns were followed (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`).
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is correct English and approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I verified that this PR follows the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I verified other components that can be impacted by these changes have been tested, and I retested again (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` have been tested & I retested again)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] If a new component is created I verified that:
    - [x] A similar component doesn't exist in the codebase
    - [x] All props are defined accurately and each prop has a `/** comment above it */`
    - [x] The file is named correctly
    - [x] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [x] The only data being stored in the state is data necessary for rendering and nothing else
    - [x] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [x] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [x] All JSX used for rendering exists in the render method
    - [x] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(themeColors.componentBG`)
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.
- [x] I have checked off every checkbox in the PR reviewer checklist, including those that don't apply to this PR.

**Comment by tienifr** [comment]
@francoisl this PR is ready to merge, could you help to check and merge it? Many thanks

**Pull request merged**